[cmd/opampsupervisor] Add support for optional agent SIGHUP config reload

douglascamata opened the pull request:
#### Description

This pull request adds a configuration option for the Collector Supervisor at `agent::use_hup_restart` that can be used to control how the Supervisor restarts the agent. 

When enabled, agent restarts will be done through sending a `SIGHUP` instead of stopping and restarting the process.

While running e2e tests with HUP restart I found an interesting race condition between the Supervisor and the Collector: if the Collector receives a SIGHUP while it's not fully started yet, it might exit prematurely. This is bad because it's a scenario that could happen quite often: a whole fleet might have local + remote configuration, then when any new machine spins up (i.e. auto-scaling) it will immediately receive a remote config. Before opening an issue on the Collector to try to figure things out over there I decided to solve it doing the following:

- If the Supervisor gets a remote config it'll only send SIGHUP to the Collector if it already reported its health at least once. This seems to ensure that everything's start and good to handle the signal well. 
- In case the health didn't arrive yet I have to wait a bit (yes, I'm also sad to wrote this). I didn't know what would be a good amount of time to wait for, so I picked the bootstrap timeout and I'm open to better suggestions if you have any. So every 100ms I check the health, up to a max wait time equal to the bootstrap timeout. If no health arrived after all the wait, the whole configuration application is considered failed.

I'm also open to any ideas on what to do if the agent didn't report health yet to the Supervisor. 

#### Link to tracking issue

Fixes #40410. 

<!--Describe what testing was performed and which tests were added.-->
#### Testing

I updated a bunch of e2e tests with an extra scenario, so that they are executed with and without the SIGHUP restart logic.

<!--Describe the documentation added.-->
#### Documentation

None so far, but I plan to update the Supervisor specification here if/when we agree regarding the issue I mentioned above on the description section.

douglascamata committed:
[supervisor] Add support for optional agent SIGHUP config reload

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Enable supervisor template configs to use agent hup restart

Signed-off-by: Douglas Camata
<159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Run relevant tests agent hup restart besides normal restart

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fail hup restart is didn't get health from agent

douglascamata committed:
Add changelog entry

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Make linter happy

douglascamata committed:
Merge branch 'main' into supervisor-hup-collector

douglascamata committed:
Put back e2e go build flag

douglascamata commented:
@TylerHelmuth this is good for a review, if you got some time.

evan-bradley reviewed (commented):
Overall I like the approach here, and allowing users to take advantage of the Collector's SIGHUP feels like an improvement (though I'm not 100% sure how strong the benefits are), so thanks for taking this on.

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -1439,16 +1449,71 @@ func (s *Supervisor) runAgentProcess() {
+					gotHealth = true
+					break
+				}
+				time.Sleep(100 * time.Millisecond)
```
I understand it now. Just making it clear to avoid confusion with the bootstrap process that the Supervisor does: here the point is to wait for the Collector's startup process being finished (i.e. service is started, signal handler's set up, etc). It's difficult to know what or how long to wait for, this is way I picked the proxy logic of waiting until the 1st health report. On top of this we can't wait forever, so I decided to wait as much as we wait for the bootstrap.

Besides this clarification, I think that I can change the approach to use a channel to communicate that we got the first health report and thus the HUP can be sent safely.

douglascamata replied:
@evan-bradley took care of this in [`d3bf969` (#40522)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40522/commits/d3bf96979337187e62bc408167346d6e3be76809).

Here's a small summary of how it works:

- The `agentStarted` atomic bool works as a quick way to check whether the agent's already up and running for the HUP signal to be sent.
- `agentStartedChan` can be used as a way to wait until the agent starts if `agentStarted` is `false`. 
- The new function `waitForAgentStart` uses these two mechanisms, protecting access to `agentStartedChan` with a mutex (keep reading to learn why this is needed).
- The HUP-reload only happens after it calls `waitForAgentStart` and if the HUP signal is sent correctly it replaces the `agentStartedChan` with a new one (using the mutex to protect the operation).
- The Supervisor sets `agentStarted` to `true` and closes `agentStartedChan` when it receives a health report from the agent.

I thought about implementing this with `sync.Cond`, and I did had this implemented for a while using it. But I thought it was slightly more complicated and easier to miss-use, specially when integrating it with a timeout on the wait logic. 

Let me know what you think and thanks for the review!

douglascamata committed:
Remove unused fields in the `Supervisor` type

douglascamata committed:
Address PR review comments

douglascamata committed:
Improve names (hup restart -> hup reload)

douglascamata committed:
Protect access to `agentStartedChan` when closing it

douglascamata committed:
Fix Supervisor config generation for tests

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix e2e test

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-hup-collector

douglascamata committed:
Put back `go:build` tag in supervisor e2e test

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix and rename `Supervisor.waitForAgentStart`

The new name is `waitForAgentReady`. Other related variables were also renamed to keep it consistent.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley commented on cmd/opampsupervisor/supervisor/commander/commander.go:
```diff
@@ -95,6 +96,19 @@ func (c *Commander) Restart(ctx context.Context) error {
+	}
+
+	c.logger.Debug("Sending SIGHUP to agent process to reload config", zap.Int("pid", c.cmd.Process.Pid))
+	if err := c.cmd.Process.Signal(syscall.SIGHUP); err != nil {
```
Do you know what the behavior is if the Collector doesn't handle (or doesn't reload the config for) SIGHUP signals? It's unlikely but possible we may control a Collector that has customizations that don't match upstream's behavior.

douglascamata replied:
Correct, the reloading process becomes a no-op. The signal will be sent successfully to the agent and on the next health report it will be considered healthy and ready, even though it might still be running on an old configuration from the Supervisor's pov.

evan-bradley replied:
Great, thank you for the details.

Could you leave a comment describing the expected behavior in these three scenarios on the `(*Supervisor).hupReloadAgent` method?

evan-bradley commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -163,10 +161,15 @@ type Supervisor struct {
+	agentReadyChan chan struct{}
+	// agentReadyLock is a mutex to protect [agentStartedChan]. This is
+	// required because we need a new channel every time the agent restarts.
+	agentReadyLock *sync.Mutex
```
We've previously used a pattern where we buffer the channel with a single message then [drain it](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/opampsupervisor/e2e_test.go#L1505) before each use in case there are leftover messages. This allows us to keep the behavior of `close` where we don't block on the channel's sink while also allowing us to keep the same channel instance for each use. Would that allow us to get rid of this mutex?

evan-bradley replied:
Thanks for the detailed response. I think my concern in this case is that if we have two Goroutines waiting on the "ready" signal, we're in a roughly equally-bad place whether both goroutines handle the message or if one is canceled (I don't think any listeners will get stuck waiting since we usually have a channel read timeout). If we're expecting only a single listener, I think we risk getting into a weird state if we enter a place where there are two listeners.

I can see that we may need to broadcast agent readiness at some point, but right now I think we could simplify a bit if we had this work with 1-1 communication.

evan-bradley reviewed (commented):
Overall I like the way this looks with the channel, thanks for making that adaptation. Just a few more questions on the implementation.

douglascamata committed:
Simplify approach for HUP reload implementation

Using a 1:1 communication channel instead of a more complex 1:N.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley reviewed (commented):
This is looking pretty good, thanks for your patience and attention to detail while dealing with these.

I like the `markAgentReady` and `resetAgentReady` methods, I think those will make it easier to switch to a broadcast model in the future should we find a need.

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -1439,16 +1451,107 @@ func (s *Supervisor) runAgentProcess() {
+	bootstrapTimer := time.NewTimer(s.config.Agent.BootstrapTimeout)
+	defer bootstrapTimer.Stop()
+
+	for {
```
Good point. That's a leftover from the previous implementation. We definitely don't need the `for` loop here.

douglascamata committed:
Improve comments

douglascamata committed:
Remove unnecessary for loop when selecting on channels

douglascamata committed:
Fix typo to make linter happy

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-hup-collector

douglascamata committed:
Passing all e2e tests

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Delegate agent start only to `Supervisor.startAgent()`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Ensure HUP e2e tests get the HUP reload config enabled

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Remove commented code that got in by accident

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Move back nil health assignment to the top of new config handling

I thought this was required to make things work, but it isn't.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented on cmd/opampsupervisor/e2e_test.go:
```diff
@@ -1839,142 +1990,165 @@ func TestSupervisorLogging(t *testing.T) {
+				"config_apply_timeout": "3s",
+			}
+			if mode.UseHUPConfigReload {
+				extraConfigData["hupRestart"] = "true"
```
Wow, I can't believe I let this pass :(

To ensure this doesn't happen anymore, I added an extra check to HUP-enabled tests that ensure the parsed Supervisor configuration has the respective config enabled (see [`52644bc` (#40522)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40522/commits/52644bc461649a858ae67a9ff26de423d2e42798)).

After I fixed this I found that some tests were failing. Now they are also fixed.

evan-bradley reviewed (approved):
Thanks @douglascamata!

One thing I want to call out so it is written somewhere is that it's still unclear how much practical benefit SIGHUP reloading brings vs. reloading the whole process at this point, despite being conceptually cleaner. I haven't profiled it myself, but I suspect that starting pipelines comprises the majority of the Collector's startup time before it is ready to receive data, and that SIGHUP reloads may not be substantially faster. The biggest benefit I can think of is that confmap Providers with persistent connections will likely stay alive through the reload. I think we'll likely want to consider benchmarking and documenting this in the future, especially if we want to consider enabling this by default.

All that said, this seems like something we want, and the functionality required to support this is sufficiently encapsulated that I'm alright including it.

evan-bradley committed:
Update cmd/opampsupervisor/supervisor/supervisor.go

evan-bradley committed:
Merge branch 'main' into supervisor-hup-collector

nenadnoveljic commented on cmd/opampsupervisor/supervisor/commander/commander.go:
```diff
@@ -95,6 +96,19 @@ func (c *Commander) Restart(ctx context.Context) error {
+		return errors.New("agent process is not running")
+	}
+
+	c.logger.Debug("Sending SIGHUP to agent process to reload config", zap.Int("pid", c.cmd.Process.Pid))
```
`SIGHUP` isn't supported on Windows. Did this change break `[e2e-tests-windows / supervisor-test](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/16030957451/job/45232053818?pr=41032)`?
`{"error": "failed to send SIGHUP to agent process: not supported by windows"}`

nenadnoveljic replied:
The PR was merged despite the error. Should changes in this module trigger [e2e-tests-windows / supervisor-test](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/16030957451/job/45232053818?pr=41032)?

nenadnoveljic replied:
[This element](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40522/files#diff-239707dd27c36360814b762f18855fdb907bca784df484a72e90d6106daf3de6R231-R234) can be made conditional to skip `SIGHUP` tests on Windows.

douglascamata replied:
Will take care of this tomorrow as early as possible. 👍

douglascamata replied:
@nenadnoveljic I think your link doesn't work well. Maybe it requires certain files to be expanded for the anchor to work? Could you share your suggestion in a different way?

nenadnoveljic replied:
Fixing the tests [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41070)
